Desktop: Resolves #4614: Allow registering multiple settings in one call

jalajcodes opened the pull request:
- Added new method `joplin.settings.registerSettings`
- Marked `joplin.settings.registerSetting` as deprecated
- Modified `joplin.settings.registerSetting` to use `joplin.settings.registerSettings`
- Added test for the new method.
Resolves #4614

roman-r-m commented on packages/lib/services/plugins/api/JoplinSettings.ts:
```diff
@@ -39,32 +39,45 @@ export default class JoplinSettings {
-		if ('minimum' in settingItem) internalSettingItem.minimum = settingItem.minimum;
-		if ('maximum' in settingItem) internalSettingItem.maximum = settingItem.maximum;
-		if ('step' in settingItem) internalSettingItem.step = settingItem.step;
+			await Setting.registerSetting(this.namespacedKey(key), internalSettingItem);
```
Maybe it's better to await only once using `Promise.all`? Should be faster when registering multiple settings.

jalajcodes replied:
Something like this? https://gist.github.com/jalajcodes/40e8e56fa1aeed1ef3afeabe91e60537

roman-r-m replied:
`return await Promise.all(settingsPromises);` don't need `await` here, let the caller await when needed.

jalajcodes replied:
Yeah, you're right. I'll push the changes.

roman-r-m commented:
You also need to add your changes to `packages/generator-joplin`, maybe in some other places too.
Then check if there are any tests for the old method then it'd be a good idea to add tests for your new one too.

laurent22 commented:
@jalajcodes, regarding the `Promise.all` change, it obviously made the code more complex and perhaps could have some issues since you are concurrently making multiple insert operations. So there are possible drawbacks to it. When you made the change, did you verify that it actually made things significantly faster? Perhaps it's even slower for all we know? So if you haven't benchmarked the change yet please do it now, and let us know the results.

As a general rule, we don't optimise unless we have a benchmark to justify the optimisation.

laurent22 commented on packages/lib/services/plugins/api/JoplinSettings.ts:
```diff
@@ -39,32 +39,46 @@ export default class JoplinSettings {
-		if ('step' in settingItem) internalSettingItem.step = settingItem.step;
+			settingsPromises.push(Setting.registerSetting(this.namespacedKey(key), internalSettingItem));
+		}
+		return Promise.all(settingsPromises);
```
What data are you returning here? The function signature does not have any return type.

laurent22 commented on packages/lib/services/plugins/api/JoplinSettings.ts:
```diff
@@ -39,32 +39,46 @@ export default class JoplinSettings {
+	 * Registers a new setting.
+	 */
+	public async registerSetting(key: string, settingItem: SettingItem) {
+		await this.registerSettings({ [key]: settingItem });
```
Please also use `this.plugin.deprecationNotice` to display a warning in the console.

jalajcodes replied:
What should be the value of `goneInVersion` inside `this.plugin.deprecationNotice()` ? 
Other places this is used, the value is either 1.5 or 1.8

laurent22 replied:
Yes this is fine.

jalajcodes commented:
> You also need to add your changes to `packages/generator-joplin`, maybe in some other places too.
> Then check if there are any tests for the old method then it'd be a good idea to add tests for your new one too.

I think I need to edit `JoplinSettings.ts` files inside `api` directory of `generator-joplin` as well as for all other plugins inside `packages/app-cli/test/support/plugins` 

I would have done it the first time but these `api` directories are excluded in the [workspace file](https://github.com/laurent22/joplin/blob/dev/joplin.code-workspace#L208) so they weren't showing up in my editor, which makes me wonder why it is done and if it would be okay if I edited them directly?

jalajcodes commented:
> When you made the change, did you verify that it actually made things significantly faster?

I just verified that it works, not if it made things faster. 😅

> So if you haven't benchmarked the change yet please do it now, and let us know the results.

I modified the plugin to add 50 settings and noted the time taken.
I ran the application 10 times with `Promise.all()` and 10 times without it.
Here are the results...https://gist.github.com/jalajcodes/81138432b1f875467fef2565b96fa758

> And please be critical of what we might say here

Thanks, I'll keep that in mind. 🙂

laurent22 commented:
@jalajcodes, thanks for looking into it.  What's your interpretation of the numbers?

jalajcodes commented:
> @jalajcodes, thanks for looking into it. What's your interpretation of the numbers?

as the numbers are very close to each other, I'd say it's fine to go either way.

but if we take into consideration that `Promise.all()` makes code complex and the concurrent insertions may cause issues later then I think we should revert back to the previous code I wrote. Anyhow, I don't have much experience so I want to know what you folks think?

jalajcodes commented:
> > You also need to add your changes to `packages/generator-joplin`, maybe in some other places too.
> > Then check if there are any tests for the old method then it'd be a good idea to add tests for your new one too.
> 
> I think I need to edit `JoplinSettings.ts` files inside `api` directory of `generator-joplin` as well as for all other plugins inside `packages/app-cli/test/support/plugins`
> 
> I would have done it the first time but these `api` directories are excluded in the [workspace file](https://github.com/laurent22/joplin/blob/dev/joplin.code-workspace#L208) so they weren't showing up in my editor, which makes me wonder why it is done and if it would be okay if I edit the files directly?

@laurent22 any info about that? should I just edit them directly?

roman-r-m commented:
> as the numbers are very close to each other

Going back to this - when you give test results like these, you always should also give mean and std deviation at least.

jalajcodes commented:
@roman-r-m @laurent22 I updated the relevant files and added the tests. Please review again.

roman-r-m commented on packages/lib/services/plugins/api/JoplinSettings.ts:
```diff
@@ -39,32 +39,48 @@ export default class JoplinSettings {
+
+			settingsPromises.push(Setting.registerSetting(this.namespacedKey(key), internalSettingItem));
+		}
+		return Promise.all(settingsPromises);
```
Weren't you going to get rid of Promise.all?

jalajcodes replied:
the standard deviation of the code using Promise.all is lower (https://gist.github.com/jalajcodes/81138432b1f875467fef2565b96fa758)
Isn't that better?

roman-r-m commented:
1. You still have not addressed Laurent's comment about return value.
2. Please avoid force push - makes it hard to keep track of what's been reviewed already.
3. Please do not mention reviewers (with @), especially not 10 minutes after making changes. This is excessive as GitHub notifies everyone of any activity on the issue.

jalajcodes commented:
> 1. You still have not addressed Laurent's comment about return value.
> 2. Please avoid force push - makes it hard to keep track of what's been reviewed already.
> 3. Please do not mention reviewers (with @), especially not 10 minutes after making changes. This is excessive as GitHub notifies everyone of any activity on the issue.

1. I could swear I changed it😉 anyways I'll do it now.
2. I squashed my commits into one so I had to force push. Is there any way to go around that?
3. Sorry, I understand and I'll keep that in mind from now on.

roman-r-m commented:
> I squashed my commits into one so I had to force push. Is there any way to go around that?

Why is it a problem if some commits are not squashed? I believe Laurent squashes everything when merging anyway.

jalajcodes commented:
> > I squashed my commits into one so I had to force push. Is there any way to go around that?
> 
> Why is it a problem if some commits are not squashed? I believe Laurent squashes everything when merging anyway.

Not a problem per se just that the last org I was contributing to required us to squash commits so...

laurent22 commented:
- You have added many unrelated changes to your pull request, please fix it.
- Please revert the Promise.all change

laurent22 commented on packages/app-cli/tests/services/plugins/api/JoplinSettings.ts:
```diff
@@ -35,6 +35,21 @@ describe('JoplinSettings', () => {
 						label: 'My Custom Setting 2',
 					});
 
+					await joplin.settings.registerSettings({
+						'myCustomSetting3': {
+							value: 3,
+							type: 1,
+							public: true,
+							label: 'My Custom Setting 3',
+						},
+						'myCustomSetting4': {
+							value: 4,
+							type: 1,
+							public: true,
+							label: 'My Custom Setting 4',
+						}
+					})
+
```
why did you put that here? Did you check the title of the test, and do you think it's related to your change? So please either find a test related to your changes, or create a new one.

jalajcodes replied:
This is the only test I found that used `joplin.settings.registerSetting()` and as it is supposed to be replaced by `registerSettings()`, I added this here. I didn't removed the calls to `registerSetting()` though, I'll do that.

jalajcodes replied:
> why did you put that here? Did you check the title of the test, and do you think it's related to your change? So please either find a test related to your changes, or create a new one.

This is the only test I found that used joplin.settings.registerSetting() and as it is supposed to be replaced by registerSettings(), I added this here. I didn't removed the calls to registerSetting() though, I'll do that.

laurent22 commented on packages/app-cli/tests/services/plugins/api/JoplinSettings.ts:
```diff
@@ -35,6 +35,21 @@ describe('JoplinSettings', () => {
 						label: 'My Custom Setting 2',
 					});
```
Since `registerSetting` is now deprecated, please update these calls.

laurent22 commented on packages/app-cli/tests/services/plugins/api/JoplinSettings.ts:
```diff
@@ -66,4 +67,38 @@ describe('JoplinSettings', () => {
+		const plugin = await service.loadPluginFromJsBundle('', pluginScript);
+		await service.runPlugin(plugin);
+
+		await expectNotThrow(() => Setting.value('plugin-org.joplinapp.plugins.PluginTest.myCustomSetting1'));
+		await expectNotThrow(() => Setting.value('plugin-org.joplinapp.plugins.PluginTest.myCustomSetting2'));
```
remove these two lines please, as the lines below already test that the call doesn't fail.

laurent22 commented on packages/generator-joplin/generators/app/templates/api/JoplinSettings.d.ts:
```diff
@@ -22,11 +22,18 @@ export default class JoplinSettings {
+	 * @deprecated Use joplin.settings.registerSettings()
+	 *
+	 * Registers a new setting.
+	 */
```
I think you need to remove this file

roman-r-m commented:
Looks good. Tested with a basic plugin, works as expected.

laurent22 commented:
Looks good now, thank you.
